src/ac_training_lab/cobot280pi/server.py

```diff
@@ -0,0 +1,178 @@
+    try:
+        cobot.send_coords(**args)
+        return {"success": True}
+    except Exception as e:
+        logger.critical(f"control coords error: {str(e)}")
+        return {"success": False, "error_msg": str(e)}
+
+def handle_query_angles(args, cobot):
+    logger.info(f"running command query/angle with {args}")
+    try:
+        angles = cobot.get_angles()
+        if angles is None or len(angles) < 6:
+            raise Exception("could not read angle")
+        return {"success": True, "angles": angles}
+    except Exception as e:
+        logger.critical(f"query angle error: {str(e)}")
+        return {"success": False, "error_msg": str(e)}
+
+def handle_query_coords(args, cobot):
+    logger.info(f"running command query/coord with {args}")
+    try:
+        coords = cobot.get_coords()
+        if coords is None or len(coords) < 6:
+            raise Exception("could not read coord")
+        return {"success": True, "coords": coords}
+    except Exception as e:
+        logger.critical(f"query coord error: {str(e)}")
+        return {"success": False, "error_msg": str(e)}
+
+def handle_query_gripper(args, cobot):
+    logger.info(f"running command query/coord with {args}")
+    try:
+        gripper_pos = cobot.get_gripper_value()
+        return {"success": True, "position": gripper_pos}
+    except Exception as e:
+        logger.critical(f"query gripper error: {str(e)}")
+        return {"success": False, "error_msg": str(e)}
+
+def handle_query_camera(args):
+    logger.info(f"running command query/camera with {args}")
+    try:
+        webcam = cv2.VideoCapture(0)
+        _, frame = webcam.read()
+        webcam.release()
+
+        img = Image.fromarray(cv2.cvtColor(frame, cv2.COLOR_BGR2RGB))
+        compressed_bytes = io.BytesIO()
+        img.save(compressed_bytes, format="JPEG", quality=args["quality"])
+        compressed_bytes.seek(0)
+        byte_str = base64.b64encode(compressed_bytes.read()).decode("utf-8")
+        return {"success": True, "image": byte_str}
+    except Exception as e:
+        logger.critical(f"query camera error: {str(e)}")
+        return {"success": False, "error_msg": str(e)}
+
+
+# MQTT Functions
+def on_connect(client, userdata, flags, rc, properties=None):
+    logger.info("Connection received with code %s." % rc)
+
+def on_publish(client, userdata, mid, properties=None):
+    logger.info("Successful publish.")
+
+def handle_message(msg, cobot):
+    # Parse payload to json dict
+    try:
+        payload_dict = json.loads(msg.payload)
+    except Exception as e:
+        return {
+            "success": False,
+            "error": str(e)
+        }
+
+    if "command" not in payload_dict:
+        return {
+            "success": False,
+            "error": "'command' key should be in payload"
+        }
+
+    # Match to a command function
+    cmd = payload_dict["command"]
+    if cmd == "control/angles":
+        return handle_control_angles(payload_dict["args"], cobot)
```

Comment:
Any function-specific restrictions on "args" that you think should be implemented?

Reply:
the only restrictions I can think of would be on control/coords and control/angles to prevent the cobot from reaching below the ground plane.

As for other restrictions, since I directly pass args to the pymycobot api with **args and catch exceptions, other error cases are directly handled by pymycobot. Let me know if you think wrapping around their functions and implementing our own checks would be better.

Reply:
I think how you implemented it is great. I'm glad you brought that up - I forgot about the ground plane. There's also potential to put the robot on a pedestal. Might be kind of interesting

Reply:
Potential pedestal - but still need to check if the clamp is wide enough to go onto both desk types in the office room.

Vondynote Set of 2 Studio Monitor Stands Desktop Clamp Speaker Stands Against The Wall https://a.co/d/1Z8Al4T